idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/components/KtFirReferenceShortener.kt

```diff
@@ -241,26 +299,38 @@ private class ElementsToShortenCollector(private val shorteningContext: FirShort
             // if qualifier is null, then this type have no package and thus cannot be shortened
             if (typeElement.qualifier == null) return null
 
-            val firstFoundClass = shorteningContext.findFirstClassifierInScopesByName(positionScopes, classId.shortClassName)?.classId
-
-            if (firstFoundClass == classId) {
-                return ShortenType(typeElement)
-            }
-        }
-
-        // none class matched
-        val (mostTopLevelClassId, mostTopLevelTypeElement) = allClassIds.zip(allTypeElements).last()
-        val availableClassifier = shorteningContext.findFirstClassifierInScopesByName(positionScopes, mostTopLevelClassId.shortClassName)
+            val option = classShortenOption(
+                shorteningContext.firSession.symbolProvider.getClassLikeSymbolByFqName(classId) as? FirRegularClassSymbol ?: return null
+            )
+            if (option == ShortenOption.DO_NOT_SHORTEN) continue
 
-        check(availableClassifier?.classId != mostTopLevelClassId) {
-            "If this condition were true, we would have exited from the loop on the last iteration. ClassId = $mostTopLevelClassId"
-        }
+            val availableClassifier = shorteningContext.findFirstClassifierInScopesByName(positionScopes, classId.shortClassName)
 
-        return if (availableClassifier == null || availableClassifier.isFromStarOrPackageImport) {
-            ShortenType(mostTopLevelTypeElement, mostTopLevelClassId.asSingleFqName())
-        } else {
-            findFakePackageToShorten(mostTopLevelTypeElement)
+            when {
+                availableClassifier == null -> {
+                    if (option == ShortenOption.SHORTEN_IF_ALREADY_IMPORTED) continue
```

Comment:
`availableClassifier == null` implies the symbol is not imported in the current file. The option `SHORTEN_IF_ALREADY_IMPORTED` means only shorten if the symbol is already imported. 

`availableClassifier.symbol != classId` when there are two classes in different pacakges with the same name and one is imported while the other is fully-qualified. In that case, in general we should not import the fully qualified symbol due to possible conflicts. But, we can still import if
* the already imported symbol is not used (this is currently not checked, and hence TODO at line 87), AND
* the already imported symbol is available from star import or from the package. In this case,
  * if the symbol is imported by default, then a star import can "overwrite it"
  * otherwise, only explicit import can "overwrite it"

Reply:
Thanks for the explanation. I think some comment and/or examples on the last 2 branches in here would be helpful.

Reply:
By the way, where do you get the rules that you described of when you can still import it? Is this based on current FE 1.0 intention's behavior?

Reply:
The overwriting rules are observed from FE1.0's behavior. I didn't pay much attention to how FE1.0's intention behaves, but it seems the current FIR implementation behaves identically from my observation.